docs/design/incrementaldownload.md

```diff
@@ -0,0 +1,65 @@
+# Incremental Download
+
+## Overview
+The feature is to allow Incremental Download of the Artifact in release so that all of the artifact items are not downloaded from the source provider and can be retrieved from a pre-maintained cache as well.
+
+## Task Changes
+Ideally we should be given the hashes of the files from the service itself but since services like Jenkins and Teamcity do not provide such hashes, we needed some generic way to get the hashes.
+
+A new task will be created named `Generate Artifact Metadata` that will generate `artifact-metadata.csv` file containing the filepaths and hashes of artifact items mapped to one another, created in the same directory as that of the artifacts.
+
+Currently, it supports VSTS but similar tasks can be written for Jenkins and teamcity as well. 
+
+##### artifact-metadata.csv example
+![Alt text](./artifact-metadata(.csv).png?raw=true "Title")
+
+## Artifact Engine Changes
+A CacheProvider will be initialized with a “Hash” that reads the old artifact-metadata.csv
+
+#### PsuedoCode
+```
+  If(deploymentInput.enableIncrementalDownload is false)
+	{
+		//download all the artifacts
+		Return releasedefinition.Artifacts;
+	}
+  Else
+	{
+		//download the artifacts incrementally
+    		Initialize cache by reading the hash of artifact type.
+		If(artifactItem is in cache)
+		{
+			//download from the cache
+			Return cache.artifactItem;
+		}
+		Else
+		{
+			//download from the sourceProvider
+			Return sourceProvider.artifactItem;
+		}
+	}
+```
+
+A new cache directory will be generated by the artifact engine under the folder _work\ArtifactEngineCache\SHA256HashofArtifact\ where the hash will be calculated by the following keys corresponding to different artifact types:
+- Keys:
```

Comment:
Will this be an input to artifact-engine? In general, would be good to clarify what are the changes to artifact-engine interface.